### PR TITLE
[KYUUBI #5136][Bug] Spark App may hang forever if FinalStageResourceManager killed all executors

### PR DESCRIPTION
**Pull request by zhouyifan279**
### _Why are the changes needed?_
In minor cases, Spark Stage hangs forever when spark.sql.finalWriteStage.eagerlyKillExecutors.enabled is true.

The bug occurs if two conditions are met in the same time:

1. All executors are either removed because of idle time out or killed by FinalStageResourceManager.
   Target executor num in YarnAllocator will be set to 0 and no more executor will be launched.
2. Target executor num in ExecutorAllocationManager equals to the executor num needed by final stage.
   Then ExecutorAllocationManager will not sync target executor num to YarnAllocator.

### _How was this patch tested?_
- [x] Add a new test suite `FinalStageResourceManagerSuite`

### CODE REVIEW

File: extensions/spark/kyuubi-extension-spark-3-3/src/main/scala/org/apache/spark/sql/FinalStageResourceManager.scala [hunk 1]
```diff
@@ -192,26 +192,55 @@ case class FinalStageResourceManager(session: SparkSession)
...
+      // Ensure target executor number has been updated in cluster manager client
+      executorAllocationClient.requestExecutors(0)
+    }
+
     executorAllocationClient.killExecutors(
       executorIds = executorsToKill,
-      adjustTargetNumExecutors = true,
+      adjustTargetNumExecutors = false,
       countFailures = false,
       force = false)
```
Comment (ulysses-you) [inline extensions/spark/kyuubi-extension-spark-3-3/src/main/scala/org/apache/spark/sql/FinalStageResourceManager.scala:217]: Shall we call `client.requestTotalExecutors` to adjust the target executor if `targetExecutors` < `draTargetExecutors` after kill executors ?

We might face a issue similar with https://github.com/apache/spark/pull/19048

File: extensions/spark/kyuubi-extension-spark-3-3/src/main/scala/org/apache/spark/sql/FinalStageResourceManager.scala [hunk 2]
```diff
@@ -192,26 +192,55 @@ case class FinalStageResourceManager(session: SparkSession)
...
+        Thread.sleep(100)
+        draTargetExecutors =
+            getDraTargetExecutors(sc, executorAllocationClient).getOrElse(0)
+    }
+    logInfo(s"DRA target executor number changed to $draTargetExecutors in " +
+      s"${System.currentTimeMillis() - startTime} ms.")
+
+    if (draTargetExecutors <= targetExecutors) {
+      // Ensure target executor number has been updated in cluster manager client
+      executorAllocationClient.requestExecutors(0)
```
Comment (ulysses-you) [inline extensions/spark/kyuubi-extension-spark-3-3/src/main/scala/org/apache/spark/sql/FinalStageResourceManager.scala:210]: I'm not sure I get it correctly. It seems we should do nothing if the dra target executor is smaller than our target executor. If it happens, the executors are pending to kill.

File: extensions/spark/kyuubi-extension-spark-3-3/src/main/scala/org/apache/spark/sql/FinalStageResourceManager.scala (hunk 1 above)
Comment (ulysses-you) [reply extensions/spark/kyuubi-extension-spark-3-3/src/main/scala/org/apache/spark/sql/FinalStageResourceManager.scala:217]: IMO, we just need to tune adjustTargetNumExecutors from true to false and call client.requestTotalExecutors to ensure the final target executor is expected.

Comment (zhouyifan279) [reply extensions/spark/kyuubi-extension-spark-3-3/src/main/scala/org/apache/spark/sql/FinalStageResourceManager.scala:217]: Agree. But build parameters of `client.requestTotalExecutors` involves many details. 
So I prefer to wait `ExecutorAllocationManager` to call `client.requestTotalExecutors`.

As long as we did not call `killExecutors` with `adjustTargetNumExecutors = true`, `ExecutorAllocationManager` should be able to manage target executor num correctly.

File: .github/workflows/master.yml [hunk 3]
```diff
@@ -59,17 +59,17 @@ jobs:
           - java: 8
             spark: '3.3'
             spark-archive: '-Dspark.archive.mirror=https://archive.apache.org/dist/spark/spark-3.1.3 -Dspark.archive.name=spark-3.1.3-bin-hadoop3.2.tgz -Pzookeeper-3.6'
-            exclude-tags: '-Dmaven.plugin.scalatest.exclude.tags=org.scalatest.tags.Slow,org.apache.kyuubi.tags.DeltaTest,org.apache.kyuubi.tags.IcebergTest'
+            exclude-tags: '-Dmaven.plugin.scalatest.exclude.tags=org.scalatest.tags.Slow,org.apache.kyuubi.tags.DeltaTest,org.apache.kyuubi.tags.IcebergTest,org.apache.kyuubi.tags.SparkLocalClusterTest'
```
Comment (zhouyifan279) [inline .github/workflows/master.yml:62]: Task serialized by Spark 3.3 Driver can not be deserialized by Spark 3.1.3 Executor

**Comment by zhouyifan279** [comment]
cc @ulysses-you @pan3793 Ready for review.

File: extensions/spark/kyuubi-extension-spark-3-3/src/test/scala/org/apache/spark/sql/FinalStageResourceManagerSuite.scala [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    //    This is possible because FinalStageResourceManager retained executors may already be
+    //    requested to be killed but not died yet.
+    // 5. Final Stage required executors is 1.
+    withSQLConf(
+      (KyuubiSQLConf.FINAL_WRITE_STAGE_EAGERLY_KILL_EXECUTORS_KILL_ALL.key, "true")) {
+      withTable("final_stage") {
+        eventually(timeout(Span(10, Minutes))) {
+          sql(
+            "CREATE TABLE final_stage AS SELECT id, count(*) as num FROM (SELECT 0 id) GROUP BY id")
+        }
```
Comment (ulysses-you) [inline extensions/spark/kyuubi-extension-spark-3-3/src/test/scala/org/apache/spark/sql/FinalStageResourceManagerSuite.scala:57]: shall we `getAdjustedTargetExecutors` at the end to make sure the target executor number is 1 ?

Comment (zhouyifan279) [reply extensions/spark/kyuubi-extension-spark-3-3/src/test/scala/org/apache/spark/sql/FinalStageResourceManagerSuite.scala:57]: done

**Review by ulysses-you — APPROVED** [review]

**Comment by pan3793** [comment]
Thanks, merged to master

**Pull request closed without merging**